core/src/main/scala/kafka/server/metadata/QuotaMetadataProcessor.scala

```diff
@@ -0,0 +1,212 @@
+                             val quotaCache: QuotaCache) extends BrokerMetadataProcessor with Logging {
+
+  override def process(event: BrokerMetadataEvent): Unit = {
+    event match {
+      case MetadataLogEvent(apiMessages, _) =>
+        apiMessages.forEach {
+          case record: QuotaRecord => handleQuotaRecord(record)
+          case _ => // We only care about quota records
+        }
+      case _ => // We only about metadata events
+    }
+  }
+
+  private[server] def handleQuotaRecord(quotaRecord: QuotaRecord): Unit = {
+    val entityMap = mutable.Map[String, String]()
+    quotaRecord.entity().forEach { entityData =>
+      entityMap.put(entityData.entityType(), entityData.entityName())
+    }
+
+    if (entityMap.contains(ClientQuotaEntity.IP)) {
+      // In the IP quota manager, None is used for default entity
+      val ipEntity = Option(entityMap(ClientQuotaEntity.IP)) match {
+        case Some(ip) => IpEntity(ip)
+        case None => DefaultIpEntity
+      }
+      handleIpQuota(ipEntity, quotaRecord)
+    } else if (entityMap.contains(ClientQuotaEntity.USER) || entityMap.contains(ClientQuotaEntity.CLIENT_ID)) {
+      // Need to handle null values for default entity name, so use "getOrElse" combined with "contains" checks
+      val userVal = entityMap.getOrElse(ClientQuotaEntity.USER, null)
+      val clientIdVal = entityMap.getOrElse(ClientQuotaEntity.CLIENT_ID, null)
+
+      // In User+Client quota managers, "<default>" is used for default entity, so we need to represent all possible
+      // combinations of values, defaults, and absent entities
+      val userClientEntity = if (entityMap.contains(ClientQuotaEntity.USER) && entityMap.contains(ClientQuotaEntity.CLIENT_ID)) {
+        if (userVal == null && clientIdVal == null) {
+          DefaultUserDefaultClientIdEntity
+        } else if (userVal == null) {
+          DefaultUserClientIdEntity(clientIdVal)
+        } else if (clientIdVal == null) {
+          UserDefaultClientIdEntity(userVal)
+        } else {
+          UserClientIdEntity(userVal, clientIdVal)
+        }
+      } else if (entityMap.contains(ClientQuotaEntity.USER)) {
+        if (userVal == null) {
+          DefaultUserEntity
+        } else {
+          UserEntity(userVal)
+        }
+      } else {
+        if (clientIdVal == null) {
+          DefaultClientIdEntity
+        } else {
+          ClientIdEntity(clientIdVal)
+        }
+      }
+      handleUserClientQuota(
+        userClientEntity,
+        quotaRecord
+      )
+    } else {
+      warn(s"Ignoring unsupported quota entity ${quotaRecord.entity()}")
+    }
+  }
+
+  private[server] def handleIpQuota(ipEntity: QuotaEntity, quotaRecord: QuotaRecord): Unit = {
+    val inetAddress = ipEntity match {
+      case IpEntity(ip) => try {
+        Some(InetAddress.getByName(ip))
+      } catch {
+        case _: UnknownHostException => throw new IllegalArgumentException(s"Unable to resolve address $ip")
```

Comment:
We warn about unknown types/keys elsewhere, so maybe we should warn in this case as well?  Otherwise nothing else will get processed from this message.

Reply:
Hmm, yea. It's actually a bit dubious to try and resolve the IP when the quota is set. What if there is a network issue at that moment? We'll miss the record and never update the quota manager.